src/compiler/checker.ts

```diff
@@ -6716,12 +6727,44 @@ namespace ts {
                         !(p.flags & SymbolFlags.Prototype || p.escapedName === "prototype" || p.valueDeclaration && getEffectiveModifierFlags(p.valueDeclaration) & ModifierFlags.Static && isClassLike(p.valueDeclaration.parent));
                 }
 
+                function sanitizeJSDocImplements(clauses: readonly ExpressionWithTypeArguments[]): ExpressionWithTypeArguments[] | undefined {
+                    const result = mapDefined(clauses, e => {
+                        const oldEnclosing = context.enclosingDeclaration;
+                        context.enclosingDeclaration = e;
+                        let expr = e.expression;
+                        if (isEntityNameExpression(expr)) {

```

Comment:
why don't we track other kinds of expressions? Just because it's too complicated? (I assume we could theoretically walk the expression tree resolving entity names as we go.)

Reply:
So... The node is called `ExpressionWithTypeArguments` because for `extends` clauses, arbitrary expressions are allowed, however I'm pretty sure we restrict `implements` clauses to `EntityNameExoression`s because they're all that make sense (since we need a type reference).